lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+    return Ufe::Vector3d(value.Get<float>(), 0, 0);
+}
+
+Ufe::Vector3d fromY(const VtValue& value) {
+    return Ufe::Vector3d(0, value.Get<float>(), 0);
+}
+
+Ufe::Vector3d fromZ(const VtValue& value) {
+    return Ufe::Vector3d(0, 0, value.Get<float>());
+}
+
+CvtRotXYZFromAttrFn getCvtRotXYZFromAttrFn(const TfToken& opName)
+{
+    static std::map<TfToken, CvtRotXYZFromAttrFn> cvt = {
+        {TfToken("xformOp:rotateX"),   fromX},
+        {TfToken("xformOp:rotateY"),   fromY},
+        {TfToken("xformOp:rotateZ"),   fromZ},
+        {TfToken("xformOp:rotateXYZ"), fromXYZ},
+        {TfToken("xformOp:rotateXZY"), fromXZY},
+        {TfToken("xformOp:rotateYXZ"), fromYXZ},
+        {TfToken("xformOp:rotateYZX"), fromYZX},
+        {TfToken("xformOp:rotateZXY"), fromZXY},
+        {TfToken("xformOp:rotateZYX"), fromZYX},
+        {TfToken("xformOp:orient"),    nullptr}}; // FIXME, unsupported.
+
+    return cvt.at(opName);
+}
+
+}
+
+namespace MAYAUSD_NS_DEF {
+namespace ufe {
+
+namespace {
+
+inline Ufe::Transform3d::Ptr nextTransform3d(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item
+) {
+    return nextHandler->transform3d(item);
+}
+
+inline Ufe::Transform3d::Ptr nextEditTransform3d(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item
+) {
+    return nextHandler->editTransform3d(item);
+}
+
+typedef Ufe::Transform3d::Ptr (*NextTransform3dFn)(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item
+);
+
+Ufe::Transform3d::Ptr createTransform3d(
+    const Ufe::Transform3dHandler::Ptr& nextHandler,
+    const Ufe::SceneItem::Ptr&          item,
+    NextTransform3dFn                   nextTransform3dFn
+)
+{
+    UsdSceneItem::Ptr usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+#if !defined(NDEBUG)
+    assert(usdItem);
+#endif
+
+    // If the prim isn't transformable, can't create a Transform3d interface
+    // for it.
+    UsdGeomXformable xformSchema(usdItem->prim());
+    if (!xformSchema) {
+        return nullptr;
+    }
+    bool resetsXformStack = false;
+    auto xformOps = xformSchema.GetOrderedXformOps(&resetsXformStack);
+
+    // Early out: if there are no transform ops yet, it's a match.
+    if (xformOps.empty()) {
+        return UsdTransform3dMayaXformStack::create(usdItem);
+    }
+
+    // If the prim supports the Maya transform stack, create a Maya transform
+    // stack interface for it, otherwise delegate to the next handler in the
+    // chain of responsibility.
+    auto stackOps = UsdMayaXformStack::MayaStack().MatchingSubstack(xformOps);
+
+    return stackOps.empty() ? nextTransform3dFn(nextHandler, item) :
+        UsdTransform3dMayaXformStack::create(usdItem);
+}
+
+// Helper class to factor out common code for translate, rotate, scale
+// undoable commands.
+class UsdTRSUndoableCmdBase : public Ufe::SetVector3dUndoableCommand {
```

Comment:
Base class for all undoable commands to set translation, rotation, scale, {rotate, scale} x {pivot, pivot compensation}.